(#1541) Add priority ordering for repository optimizations

AdmiringWorm opened the pull request:
This pull requests re-adds the ability to order sources by priority when installing and upgrading packages.

fixes #1541

gep13 commented on src/chocolatey/infrastructure.app/nuget/NugetList.cs:
```diff
@@ -180,6 +187,27 @@ public static IPackage find_package(string packageName, SemanticVersion version,
+            if (priorityRepository != null)
+            {
+                // Returns an IEnumerable of Tuple
+                // Item1 = the priory
```
```suggestion
                // Item1 = the priority

```

gep13 commented on src/chocolatey/infrastructure.app/nuget/NugetList.cs:
```diff
@@ -180,6 +187,27 @@ public static IPackage find_package(string packageName, SemanticVersion version,
+            {
+                // Returns an IEnumerable of Tuple
+                // Item1 = the priory
+                // Item2 = the repository list for that priory
```
```suggestion
                // Item2 = the repository list for that priority

```

gep13 commented on src/chocolatey/infrastructure.app/nuget/PriorityAggregateRepository.cs:
```diff
@@ -0,0 +1,160 @@
+namespace chocolatey.infrastructure.app.nuget
+{
+    /// <summary>
+    /// Helper file for grouping repositories with a distinct priory set.
```
```suggestion
    /// Helper file for grouping repositories with a distinct priority set.
```

gep13 commented on src/chocolatey.tests.integration/scenarios/InstallScenarios.cs:
```diff
@@ -2992,6 +2992,154 @@ public void should_have_an_error_package_result()
+                }
+            }
+        }
+
```
There is an extra unnecessary line here.

gep13 reviewed (changes requested):
A couple minor changes based on a quick review.

gep13 commented:
@AdmiringWorm would it be possible for you to update this PR with the new template that we are using in other repositories, i.e. include information about the testing that was performed (in addition to the new tests which have been added - thanks for getting those added!)

AdmiringWorm commented:
@gep13 I have fixed the requested changes, updated the template for the PR and changed the PR to target the develop branch (instead of the old master branch).

The testing steps added was added from memory, and not the same exact steps I had done back when I opened the PR.
